src/core/net/srp_client.cpp

```diff
@@ -1500,51 +1503,86 @@ void Client::EnableAutoStartMode(AutoStartCallback aCallback, void *aContext)
-    // selected by the auto-start feature, we verify that the selected
+    serverSockAddr.Clear();
+
+    // If the SRP client is not running and auto start mode is
+    // enabled, we check if we can find any SRP server info from the
+    // Thread Network Data. If it is already and the server was chosen
+    // by the auto-start feature, then we ensure that the selected
     // server is still present in the Network Data.
+    //
+    // Two types of "DNS/SRP Service" entries can be present in
+    // Network Data, "DNS/SRP Service Anycast Address" model and
+    // "DNS/SRP Service Unicast" model. The Anycast entries are
+    // preferred over the Unicast entries.
 
     VerifyOrExit(!IsRunning() || mAutoStartDidSelectServer);
 
-    while (Get<NetworkData::Service::Manager>().GetNextSrpServerInfo(iterator, server) == kErrorNone)
+    if (Get<NetworkData::Service::Manager>().FindPreferredDnsSrpAnycastInfo(anycastInfo) == kErrorNone)
     {
-        numServers++;
-
-        // Choose a server randomly (with uniform distribution) from
-        // the list of servers. As we iterate through server entries,
-        // with probability `1/numServers`, we choose to switch the
-        // current selected server with the new entry. This approach
-        // results in a uniform/same probability of selection among
-        // all server entries.
-
-        if ((numServers == 1) || (Random::NonCrypto::GetUint16InRange(0, numServers) == 0))
+        if (IsRunning() && mAutoStartDidSelectServer && mAutoStartIsUsingAnycastAddress)
         {
-            selectedServer = server;
+            if ((mServerInstanceNumber == anycastInfo.mSerialNumber) &&
+                (GetServerAddress().GetAddress() == anycastInfo.mAnycastAddress))
+            {
+                // Client is already using the same anycast address.
+                ExitNow();
+            }
         }
 
-        if (IsRunning() && mAutoStartDidSelectServer && (GetServerAddress() == server.mSockAddr))
+        otLogInfoSrp("[client] Found anycast server %d", anycastInfo.mSerialNumber);
+
+        serverSockAddr.SetAddress(anycastInfo.mAnycastAddress);
+        serverSockAddr.SetPort(mAnycastServerPort);
+        mServerInstanceNumber = anycastInfo.mSerialNumber;
+        serverIsAnycast       = true;
+    }
+    else
+    {
+        uint16_t                                  numServers = 0;
+        NetworkData::Service::DnsSrpUnicast::Info unicastInfo;
+        NetworkData::Service::Manager::Iterator   iterator;
+
+        while (Get<NetworkData::Service::Manager>().GetNextDnsSrpUnicastInfo(iterator, unicastInfo) == kErrorNone)
         {
-            ExitNow();
+            if (IsRunning() && mAutoStartDidSelectServer && !mAutoStartIsUsingAnycastAddress &&
+                (GetServerAddress() == unicastInfo.mSockAddr))
+            {
+                ExitNow();
+            }
+
+            numServers++;
+
+            // Choose a server randomly (with uniform distribution) from
```

Comment:
This is a feature to consider maybe for future. It may require some thoughts (e.g. client may need to track a list of servers that it tried before and failed, and mechanism to clear to such a "deny list" table, etc). 

For now, we rely on the behavior for older "network-data" entries will be removed (or timed out) when the server goes away and then the clients will detect this and switch to another server (if there is any).

Reply:
I agree we can leave this for future. Another reason we may want to iterate is because a BR is incorrectly configured with an IPv6 address and/or port that does not reach an SRP server.